// ng2-carouselamos/__ivy_ngcc__/dist/ng2-carouselamos.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Ng2Carouselamos } from './components';
import * as ɵngcc0 from '@angular/core';
var Ng2CarouselamosModule = (function () {
    function Ng2CarouselamosModule() {
    }
Ng2CarouselamosModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2CarouselamosModule });
Ng2CarouselamosModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2CarouselamosModule_Factory(t) { return new (t || Ng2CarouselamosModule)(); }, imports: [[
            CommonModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2CarouselamosModule, { declarations: function () { return [Ng2Carouselamos]; }, imports: function () { return [CommonModule]; }, exports: function () { return [Ng2Carouselamos]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2CarouselamosModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                ],
                declarations: [
                    Ng2Carouselamos,
                ],
                exports: [
                    Ng2Carouselamos,
                ]
            }]
    }], function () { return []; }, null); })();
    return Ng2CarouselamosModule;
}());
export { Ng2CarouselamosModule };
/** @nocollapse */
Ng2CarouselamosModule.ctorParameters = function () { return []; };
